src/test/util/cluster_linearize.h

```diff
@@ -0,0 +1,255 @@
+    {
+        /** Construct a topological order to serialize the transactions in. */
+        std::vector<ClusterIndex> topo_order(depgraph.TxCount());
+        std::iota(topo_order.begin(), topo_order.end(), ClusterIndex{0});
+        std::sort(topo_order.begin(), topo_order.end(), [&](ClusterIndex a, ClusterIndex b) {
+            auto anc_a = depgraph.Ancestors(a).Count(), anc_b = depgraph.Ancestors(b).Count();
+            if (anc_a != anc_b) return anc_a < anc_b;
+            return a < b;
+        });
+
+        /** Which transactions the deserializer already knows when it has deserialized what has
+         *  has been serialized here so far, and in what order. */
+        std::vector<ClusterIndex> rebuilt_order;
+        rebuilt_order.reserve(depgraph.TxCount());
+
+        // Loop over the transactions in topological order.
+        for (ClusterIndex topo_idx = 0; topo_idx < topo_order.size(); ++topo_idx) {
+            /** Which depgraph index we are currently writing. */
+            ClusterIndex idx = topo_order[topo_idx];
+            // Write size, which must be larger than 0.
+            s << VARINT_MODE(depgraph.FeeRate(idx).size, VarIntMode::NONNEGATIVE_SIGNED);
+            // Write fee, encoded as an unsigned varint (odd=negative, even=non-negative).
+            s << VARINT(SignedToUnsigned(depgraph.FeeRate(idx).fee));
+            // Write dependency information.
+            SetType written_parents;
+            uint64_t diff = 0; //!< How many potential parent/child relations we have skipped over.
+            for (ClusterIndex dep_dist = 0; dep_dist < topo_idx; ++dep_dist) {
+                /** Which depgraph index we are currently considering as parent of idx. */
+                ClusterIndex dep_idx = topo_order[topo_idx - 1 - dep_dist];
+                // Ignore transactions which are already known to be ancestors.
+                if (depgraph.Descendants(dep_idx).Overlaps(written_parents)) continue;
+                if (depgraph.Ancestors(idx)[dep_idx]) {
+                    // When an actual parent is encounted, encode how many non-parents were skipped
+                    // before it.
+                    s << VARINT(diff);
+                    diff = 0;
+                    written_parents.Set(dep_idx);
+                } else {
+                    // When a non-parent is encountered, increment the skip counter.
+                    ++diff;
+                }
+            }
+            // Write position information.
+            ClusterIndex insert_distance = 0;
+            while (insert_distance < rebuilt_order.size()) {
+                // Loop to find how far from the end in rebuilt_order to insert.
+                if (idx > *(rebuilt_order.end() - 1 - insert_distance)) break;
+                ++insert_distance;
+            }
+            rebuilt_order.insert(rebuilt_order.end() - insert_distance, idx);
+            s << VARINT(diff + insert_distance);
```

Comment:
Don't go out of your way for now, this may just be a mental block for myself only. I worked out some examples myself and it works, I just can't mentally model the skipping somehow.

Reply:
yes this was very insightful, thank you